lightning/src/routing/network_graph.rs

```diff
@@ -243,6 +290,414 @@ impl<C: Deref + Sync + Send, L: Deref + Sync + Send> RoutingMessageHandler for N
+		}
+		// On subsequent message(s)...
+		else {
+			// We need to validate the sequence of the reply message is expected.
+			// Subsequent messages must set the first_blocknum to the previous
+			// message's first_blocknum plus number_of_blocks. There is discrepancy
+			// in implementation where some resume on the last sent block. We will
+			// loosen the restriction and accept either, and otherwise consider the
+			// message malformed and return an error.
+			let task_received_last_block = task_received_last_block.unwrap();
+			if msg.first_blocknum != task_received_last_block && msg.first_blocknum != task_received_last_block + 1 {
+				query_range_tasks.remove(their_node_id);
+				return Err(LightningError {
+					err: String::from("Failing reply_channel_range with invalid sequence"),
+					action: ErrorAction::IgnoreError,
+				});
+			}
+
+			// Next we check to see that we have received a realistic number of
+			// reply messages for a query. This caps the allocation exposure
+			// for short_channel_ids that will be batched and sent in query channels.
+			if task_number_of_replies + 1 > MAX_REPLY_CHANNEL_RANGE_PER_QUERY {
+				query_range_tasks.remove(their_node_id);
+				return Err(LightningError {
+					err: String::from("Failing reply_channel_range due to excessive messages"),
+					action: ErrorAction::IgnoreError,
+				});
+			}
+
+			// Capture the last_block in our task so that subsequent messages
+			// can be validated.
+			let task = query_range_tasks.get_mut(their_node_id).unwrap();
+			task.number_of_replies += 1;
+			task.received_last_block = Some(msg_last_block);
+		}
+
+		// We filter the short_channel_ids to those inside the query range.
+		// The most significant 3-bytes of the short_channel_id are the block.
+		{
+			let mut filtered_short_channel_ids: Vec<u64> = msg.short_channel_ids.clone().into_iter().filter(|short_channel_id| {
+				let block = short_channel_id >> 40;
+				return block >= query_range_tasks.get(their_node_id).unwrap().first_blocknum as u64 && block <= task_last_block as u64;
+			}).collect();
+			let task = query_range_tasks.get_mut(their_node_id).unwrap();
+			task.short_channel_ids.append(&mut filtered_short_channel_ids);
+		}
+
+		// The final message is indicated by a last_blocknum that is equal to
+		// or greater than the query's last_blocknum.
+		if msg_last_block >= task_last_block {
+			log_debug!(self.logger, "Completed query_channel_range: peer={}, first_blocknum={}, number_of_blocks={}", log_pubkey!(their_node_id), task_first_blocknum, task_number_of_blocks);
+
+			// We can now fire off a query to obtain routing messages for the
+			// accumulated short_channel_ids.
+			{
+				let task = query_range_tasks.get_mut(their_node_id).unwrap();
+				let mut short_channel_ids = Vec::new();
+				std::mem::swap(&mut short_channel_ids, &mut task.short_channel_ids);
+				self.query_short_channel_ids(their_node_id, task.chain_hash, short_channel_ids)?;
```

Comment:
Why not send the queries as we go? With that I think we can drop all of the state tracking here, which would clean up the code a ton and make reasoning about state and DoS issues much simpler.

Reply:
I don't think we can make the SCID queries entirely stateless. The spec only allows a single query of each type to be in-flight at a time. A large channel range query will stream multiple batches of SCIDs. Because we can only have a single SCID query in-flight, we need to store the subsequent batches and wait until we get an `_end` message before we can initiate another SCID query. For example:

```
|------ query_channel_range               ----->|
|<----- reply_channel_range (batch 1)     ------|
|<----- reply_channel_range (batch 2)     ------|
|<----- reply_channel_range (batch n)     ------|
|                                               |
|------ query_short_channel_ids (batch 1) ----->|
|<----- many chan_ann, etc...             ------|
|<----- reply_short_channel_ids_end       ------|
|                                               |
|------ query_short_channel_ids (batch 2) ----->|
|<----- many chan_ann, etc...             ------|
|<----- reply_short_channel_ids_end       ------|
|                                               |
|------ query_short_channel_ids (batch n) ----->|
|<----- many chan_ann, etc...             ------|
|<----- reply_short_channel_ids_end       ------|
```
So at a bare minimum we'll need to serialize and defer SCID query events some how, even if we initiate the first SCID query as soon as receive the first batch.

The path I went down allows us to handle the three use cases described above, but is overkill if we only plan to support full syncs. I'm happy to come up with a few options to simplify things and report back!

Reply:
Ugh. I wonder if we should just get the spec to relax the query limit for short_channel_ids data so that we can make all our responses completely stateless (modulo network buffers). The alternative would be to tightly limit the response set size and maybe even do requests one block at a time to make sure we can set a tight limit and not have to worry about it. I do think we can still cut back a bunch on the amount of state we track to enforce checks that aren't really DoS risks.